WIP [svc health check] allow configurable interval between checks. WIP

jeremymv2 opened the pull request:
### :nut_and_bolt: Description

Resolves https://github.com/habitat-sh/habitat/issues/5466

This change will allow users to specify a health check interval at service load time by storing a `health_check_interval_seconds` in the `Spec` file for the service such as:

```
$ sudo cat /hab/sup/default/specs/elasticsearch.spec
ident = "core/elasticsearch"
group = "default"
bldr_url = "https://bldr.habitat.sh"
channel = "stable"
topology = "standalone"
update_strategy = "none"
binds = []
binding_mode = "strict"
desired_state = "up"
health_check_interval_seconds = 35
```

Features of the PR:
- add `--health-check-interval` to `hab svc load ..`
- validation enforces a u32 value between 1 and 86400 seconds
- Service or census changes will trigger all Services to re-run Health Check within 30s regardless if their configured interval is > 30s.

**Note** I will follow up with test and doc updates, but first, I wanted to open this up for discussion before moving forward.

Related https://github.com/habitat-sh/habitat/issues/5326
Related https://github.com/habitat-sh/habitat/issues/5584

@christophermaier @mwrock @baumanj @raskchanky What are your thoughts on this implementation as a preliminary step to adding more usefulness to our Service Health Check feature? 

Signed-off-by: Jeremy J. Miller <jm@chef.io>

### :white_check_mark: Checklist
- [ ] Necessary tests added/updated?
- [ ] Necessary docs added/updated?
- [x] Code actually executed?
- [x] Vetting performed (unit tests, lint, etc.)?

thesentinels commented:
Thanks for the pull request! Here is what will happen next:
 1. Your PR will be reviewed by the maintainers
 2. If everything looks good, one of them will `approve` it, and your PR will be merged.

Thank you for contributing!

baumanj commented:
I'm planning to look at this tomorrow and hope to have feedback for you by Monday.

baumanj commented:
I re-started the AppVeyor job

jeremymv2 commented:
@cm @baumanj Thank you for the very helpful feedback. I believe I have addressed most of the items, but I'd still like to have some discussion around the finer details of this implementation.

HealthCheckInterval is now a custom core::Service type in https://github.com/habitat-sh/core/pull/95

Before I move further with refining this PR and testing edge cases, I'd like to get your feedback again on this intermediary point with the new commits I've just added to ensure we're all in agreement.  Thank you.

jeremymv2 commented on components/sup/src/manager/service/mod.rs:
```diff
@@ -841,9 +856,9 @@ impl Service {
-                    if Instant::now().duration_since(last_check) >= *HEALTH_CHECK_INTERVAL {
+            match self.next_health_check {
+                Some(next_check) => {
+                    if next_check <= Instant::now() {
```
I actually wish to call `self.run_health_check_hook()` not only if the value is `Some(check_time)` and `<= Instant::now()` but also if it is `None`.  When it's `None` that mean haven't yet run a check and need to.

I think this is a fair balance between being evocative and concise:
```
            match self.scheduled_health_check {
                Some(scheduled_check_instant) => {
                    //  Run a check if the scheduled health check Instant is on this tick or it has passed
                    if scheduled_check_instant <= Instant::now() {
                        self.run_health_check_hook();
                    }
                }
                None => self.run_health_check_hook(),
            }
```

baumanj replied:
The `None` case does make sense though. Thanks for explaining that. I think your proposal looks fine.

jeremymv2 commented:
@cm @baumanj some spare cycles popped up yesterday so I made some more progress. Please take a look at your convenience.

baumanj commented on components/sup/src/manager/service/mod.rs:
```diff
@@ -841,9 +859,48 @@ impl Service {
+            // something keeps flapping in the ring, we also check if self.special_health_check_scheduled
+            // is true and the Duration since self.last_health_check exceeds HealthCheckInterval::default()
+            // When that condition is met, we run a health check at that instant.
+
```
This logic overall is pretty gnarly. I wonder if we can simplify it by taking a step back. Does this sound correct?

1. We never want to run a health check *more* frequently than `HealthCheckInterval::default()`
2. Unless `self.scheduled_health_check` is set in the future, we want to run it now

If that's true, I think we can simplify this down to:
```rust
            match (self.last_health_check, self.scheduled_health_check) {
                (Some(last_check_instant), _)
                    if last_check_instant.elapsed() < HealthCheckInterval::default() =>
                {
                    trace!(
                        "Skipping health check; it was run only {:?} ago",
                        last_check_instant
                    );
                }
                (_, Some(scheduled_check_instant)) if scheduled_check_instant > now => {
                    trace!(
                        "Skipping health check; next scheduled for {:?} (now: {:?})",
                        scheduled_check_instant,
                        now
                    );
                }
                _ => self.run_health_check_hook(),
            }
```

I think this allows us to do away with the `special_health_check_scheduled` entirely. Am I missing something?

jeremymv2 replied:
Unfortunately I do not think I agree with premise 1. as it feels a bit inflexible with some edge-case users likely requesting more frequent checks in the future, but I can probably bend if that makes the implementation less complex. Reading 2. and thinking..

baumanj replied:
We could go with 1 for now and see if a real use case for more frequent health checks develops in the future. As always, I default to YAGNI when code is getting complex.

jeremymv2 replied:
@baumanj I believe that the following captures all the scenarios and is a good balance between clarity and conciseness. Also, I added some [comments](https://github.com/habitat-sh/habitat/pull/5890/files#diff-320293bb5f68f709add66ac43f5ea52eR863) that I think summarize the implementation as a whole.

```
            match (self.scheduled_health_check, self.last_health_check) {
                (Some(scheduled_check_instant), _) => {
                    // the check time has passed or is now, so we run a check
                    if scheduled_check_instant <= now {
                        self.run_health_check_hook();
                    }
                }
                (_, Some(last_check_instant)) => {
                    // we scheduled a special check so we will honor that contract
                    if self.special_health_check_scheduled
                        && last_check_instant.elapsed() >= *HealthCheckInterval::default().as_ref()
                    {
                        self.run_health_check_hook();
                    }
                }
                _  => self.run_health_check_hook(),
            }
```

baumanj replied:
I'm not really clear why we need the `special_health_check_scheduled` Boolean. It seems to add noticeable complexity, but I'm not seeing the use case that makes it necessary. What functionality do we need _right now_ that's not captured by [these two simple rules](https://github.com/habitat-sh/habitat/pull/5890#discussion_r242289186)?

If we can use that, we have only 3 cases:
1. Don't run the check because it's too frequent
1. Don't run the check because it's not time yet
1. Run the check

With the code you're proposing here there are 5 cases:
1. Run the check because it's at or past the scheduled time
1. Don't run the check because it's before the scheduled time
1. Run the check because there's a special check scheduled and it it's past the default interval since the last check
1. Don't run the check because there's either not a special check scheduled or it's too frequent
1. Run the check

Also, the 3-case version has the benefit of covering all the reasons we *don't* want to run the check first and then letting the catch-all be running the check. The 5-case version has nested conditions which go back and forth between running and not running, which (at least for me) make it harder to reason about.

That's a lot more cognitive overhead, I think. Plus the maintenance of the `special_health_check_scheduled` Boolean through other parts of the code.

jeremymv2 replied:
I believe there is one case that you are not accounting for and is why we need the `special_health_check_scheduled` boolean. 

Since we schedule an 'out of band' health check when something changes in the Supervisor mesh, we need to mitigate a situation where we could perpetually extend the `scheduled_health_check` because the thing is perpetually flapping.  When we schedule an 'out of band' health check we need to guarantee that it will run within `HealthCheckInterval::default()`.

If we mark the fact that a `self.special_health_check_scheduled` is true, then we know that we HAVE to run a new check within `HealthCheckInterval::default()` interval - regardless of what the actual configured interval is.

jeremymv2 replied:
@baumanj I think moving the perpetual re-scheduling mitigation into the `schedule_health_check()` function is prudent. Please take a look at my last commit. I think it makes use of your recommendations. Thank you!

jeremymv2 replied:
It does require [this](https://github.com/habitat-sh/habitat/pull/5890/files#diff-320293bb5f68f709add66ac43f5ea52eR951)

baumanj replied:
I should be able to look at this tomorrow. Thanks for being so open to feedback, as always!

baumanj commented on components/sup/src/manager/service/mod.rs:
```diff
@@ -926,10 +986,46 @@ impl Service {
+                "Service ({}) health check is: {}. Re-running within default {:#?}",
+                self.spec_ident,
+                check_result,
+                *HealthCheckInterval::default().as_ref()
```
```suggestion
                 HealthCheckInterval::default()
```

jeremymv2 replied:
Yes, good catch! Now that I implemented `fmt::Display for HealthCheckInterval` that will work and is easier on the eyes.

baumanj commented on components/sup/src/manager/service/mod.rs:
```diff
@@ -926,10 +986,46 @@ impl Service {
     }
 
+    fn schedule_routine_health_check(&mut self) {
+        let interval = self.health_check_interval.clone();
```
```rust
        self.schedule_health_check(self.health_check_interval);
```
won't work because of:
```
error[E0503]: cannot use `self.health_check_interval` because it was mutably borrowed                                                                                                                       
    --> components/sup/src/manager/service/mod.rs:1031:36                                                                                                                                                   
     |                                                                                                                                                                                                      
1031 |         self.schedule_health_check(self.health_check_interval);                                                                                                                                      
     |         ----                       ^^^^^^^^^^^^^^^^^^^^^^^^^^ use of borrowed `*self`                                                                                                                
     |         |                                                                                                                                                                                            
     |         borrow of `*self` occurs here                                                                                                                                                                
```
But since `HealthCheckInteval` is `Copy` type, there's not need to `clone`:
```suggestion
        let interval = self.health_check_interval;
```

jeremymv2 replied:
It's really odd that it compiles as-is for me 🤔 but I agree with your comments here and am taking your suggestion also yields a successful compile.

baumanj replied:
It's not surprising that it compiles (everything that's `Copy` also implements `Clone`), but since `clone` typically implies we're doing something expensive, avoiding calling it on a `Copy` type helps prevent unnecessary head-scratching.

jeremymv2 replied:
Thank you for the clarification - makes perfect sense.

baumanj commented on components/sup/src/manager/service/mod.rs:
```diff
@@ -928,10 +945,59 @@ impl Service {
+                self.spec_ident,
+                check_result,
+                HealthCheckInterval::default()
+            );
```
As is, the logging would look like
```
DEBUG 2018-12-19T15:46:51Z: habitat_sup::manager::service: Scheduling next health check for (foo) in (30s)
DEBUG 2018-12-19T15:46:51Z: habitat_sup::manager::service: Service (foo) health check is: CRITICAL. Re-running within default HealthCheckInterval(
    30s
)
```
I'd suggest moving the `debug!` here above the call to `self.schedule_special_health_check` and changing it to:
```rust
            debug!(
                "Service ({}) health check is: {}; scheduling special health check",
                self.spec_ident, check_result
            );
```
giving nicer logging:
```
DEBUG 2018-12-19T15:46:51Z: habitat_sup::manager::service: Service (foo) health check is: CRITICAL; scheduling special health check
DEBUG 2018-12-19T15:46:51Z: habitat_sup::manager::service: Scheduling next health check for (foo) in (30s)
```

jeremymv2 replied:
That works for me! See this last [commit](https://github.com/habitat-sh/habitat/pull/5890/commits/0b33755fd040ec33da4b2e7c85e8faa0d4bd6436)

jeremymv2 commented:
let's kick off a rebuild now since the PR to [core](https://github.com/habitat-sh/core/pull/95) merged.

jeremymv2 commented:
Once this merges, I will follow up with a [doc](https://github.com/habitat-sh/habitat/issues/5466) update.

baumanj commented:
Hey @jeremymv2, now that https://github.com/habitat-sh/core/pull/95 is merged, I'm updating the `habitat` repo to include it: https://github.com/habitat-sh/habitat/pull/6009.

Once that merges, can you rebase this so that the CI passes? Then I can merge this one.

baumanj commented:
https://github.com/habitat-sh/habitat/pull/6009 is merged now

jeremymv2 commented:
I've rebased; however I think we need to attempt a restart of the Appveyor tests.

baumanj commented:
I've restarted the AppVeyor stuff. We should really get you access to do that yourself, @jeremymv2. You've certainly made a lot of valuable contributions.

baumanj commented:
/me spins the AppVeyor wheel of fortune again

baumanj commented:
AppVeyor timed out again. It was the packaging check, not the unit tests, so I'll force merge if it doesn't finish after one more try.
